File: packages/SystemUI/src/com/android/systemui/ambientmusic/AmbientIndicationService.java
package com.android.systemui.ambientmusic;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Handler;
import android.os.SystemClock;
import android.os.UserHandle;
import android.util.Log;
import com.android.keyguard.KeyguardUpdateMonitor;
import com.android.keyguard.KeyguardUpdateMonitorCallback;
import com.android.systemui.Dependency;
import java.util.Objects;

public class AmbientIndicationService extends BroadcastReceiver {
    private final AlarmManager mAlarmManager;
    private final AmbientIndicationContainer mAmbientIndicationContainer;
    private final KeyguardUpdateMonitorCallback mCallback = new KeyguardUpdateMonitorCallback() {
        public void onUserSwitchComplete(int i) {
            onUserSwitched();
        }
    };
    private final Context mContext;
    private final AlarmManager.OnAlarmListener mHideIndicationListener;

    public AmbientIndicationService(Context context, AmbientIndicationContainer ambientIndicationContainer) {
        mContext = context;
        mAmbientIndicationContainer = ambientIndicationContainer;
        mAlarmManager = (AlarmManager) context.getSystemService(AlarmManager.class);
        AmbientIndicationContainer ambientIndicationContainer2 = mAmbientIndicationContainer;
        Objects.requireNonNull(ambientIndicationContainer2);
        mHideIndicationListener = new AlarmManager.OnAlarmListener() {
            public final void onAlarm() {
                mAmbientIndicationContainer.hideAmbientMusic();
            }
        };
        start();
    }

    public void start() {
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction("com.google.android.ambientindication.action.AMBIENT_INDICATION_SHOW");
        intentFilter.addAction("com.google.android.ambientindication.action.AMBIENT_INDICATION_HIDE");
        mContext.registerReceiverAsUser(this, UserHandle.ALL, intentFilter, "com.google.android.ambientindication.permission.AMBIENT_INDICATION", (Handler) null);
        ((KeyguardUpdateMonitor) Dependency.get(KeyguardUpdateMonitor.class)).registerCallback(mCallback);
    }

    public void onReceive(Context context, Intent intent) {
        if (!isForCurrentUser()) {
            Log.i("AmbientIndication", "Suppressing ambient, not for this user.");
        } else if (verifyAmbientApiVersion(intent)) {
            if (mAmbientIndicationContainer.isMediaPlaying()) {
                Log.i("AmbientIndication", "Suppressing ambient intent due to media playback.");
                return;
            }
            String action = intent.getAction();
            char c = 65535;
            int hashCode = action.hashCode();
            if (hashCode != -1032272569) {
                if (hashCode == -1031945470 && action.equals("com.google.android.ambientindication.action.AMBIENT_INDICATION_SHOW")) {
                    c = 0;
                }
            } else if (action.equals("com.google.android.ambientindication.action.AMBIENT_INDICATION_HIDE")) {
                c = 1;
            }
            if (c == 0) {
                long min = Math.min(Math.max(intent.getLongExtra("com.google.android.ambientindication.extra.TTL_MILLIS", 180000), 0), 180000);
                mAmbientIndicationContainer.setAmbientMusic(intent.getCharSequenceExtra("com.google.android.ambientindication.extra.TEXT"), (PendingIntent) intent.getParcelableExtra("com.google.android.ambientindication.extra.OPEN_INTENT"), intent.getBooleanExtra("com.google.android.ambientindication.extra.SKIP_UNLOCK", false));
                mAlarmManager.setExact(2, SystemClock.elapsedRealtime() + min, "AmbientIndication", mHideIndicationListener, (Handler) null);
                Log.i("AmbientIndication", "Showing ambient indication.");
            } else if (c == 1) {
                mAlarmManager.cancel(mHideIndicationListener);
                mAmbientIndicationContainer.hideAmbientMusic();
                Log.i("AmbientIndication", "Hiding ambient indication.");
            }
        }
    }

    private boolean verifyAmbientApiVersion(Intent intent) {
        int intExtra = intent.getIntExtra("com.google.android.ambientindication.extra.VERSION", 0);
        if (intExtra == 1) {
            return true;
        }
        Log.e("AmbientIndication", "AmbientIndicationApi.EXTRA_VERSION is " + 1 + ", but received an intent with version " + intExtra + ", dropping intent.");
        return false;
    }

    public boolean isForCurrentUser() {
        return getSendingUserId() == getCurrentUser() || getSendingUserId() == -1;
    }

    public int getCurrentUser() {
        return KeyguardUpdateMonitor.getCurrentUser();
    }

    public void onUserSwitched() {
        mAmbientIndicationContainer.hideAmbientMusic();
    }
}

File: packages/SystemUI/src/com/google/android/systemui/ambientmusic/AmbientIndicationService.kt
package com.google.android.systemui.ambientmusic

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.SystemClock
import android.text.TextUtils
import android.util.Log
import com.android.keyguard.KeyguardUpdateMonitor
import com.android.keyguard.KeyguardUpdateMonitorCallback
import com.android.systemui.user.domain.interactor.SelectedUserInteractor
import com.google.android.systemui.keyguard.domain.interactor.AmbientIndicationInteractor
import com.google.android.systemui.keyguard.shared.AmbientIndicationMusicStatus
import com.google.android.systemui.keyguard.shared.ExpandedIndicationData
import com.google.android.systemui.keyguard.shared.ExtendedIndication

class AmbientIndicationService(
    private val alarmManager: AlarmManager,
    private val context: Context,
    private val keyguardUpdateMonitor: KeyguardUpdateMonitor,
    private val selectedUserInteractor: SelectedUserInteractor,
    private val ambientIndicationInteractor: AmbientIndicationInteractor,
) : BroadcastReceiver() {

    private val TAG = "AmbientIndication"

    var mStarted: Boolean = false

    internal val mCallback: KeyguardUpdateMonitorCallback =
        object : KeyguardUpdateMonitorCallback() {
            override fun onUserSwitchComplete(userId: Int) {
                onUserSwitched()
            }
        }

    private val mHideIndicationListener: AlarmManager.OnAlarmListener =
        AlarmManager.OnAlarmListener {
            ambientIndicationInteractor.hideAmbientMusic()
        }

    fun getCurrentUser(): Int = selectedUserInteractor.getSelectedUserId()

    fun isForCurrentUser(): Boolean {
        return sendingUserId == getCurrentUser() || sendingUserId == -1
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (!isForCurrentUser()) {
            Log.i(TAG, "Suppressing ambient, not for this user.")
            return
        }

        val version = intent.getIntExtra("com.google.android.ambientindication.extra.VERSION", 0)
        if (version != 1) {
            Log.e(
                TAG,
                "AmbientIndicationApi.EXTRA_VERSION is 1, but received an intent with version $version, dropping intent.",
            )
            return
        }

        val text = intent.getCharSequenceExtra("com.google.android.ambientindication.extra.TEXT")
        val openIntent =
            intent.getParcelableExtra<PendingIntent>(
                "com.google.android.ambientindication.extra.OPEN_INTENT"
            )
        val favoritingIntent =
            intent.getParcelableExtra<PendingIntent>(
                "com.google.android.ambientindication.extra.FAVORITING_INTENT"
            )
        val songTitle =
            intent.getCharSequenceExtra("com.google.android.ambientindication.extra.SONG_TITLE")
        val artistName =
            intent.getCharSequenceExtra("com.google.android.ambientindication.extra.ARTIST_NAME")
        val ttlMillis =
            Math.min(
                Math.max(
                    intent.getLongExtra(
                        "com.google.android.ambientindication.extra.TTL_MILLIS",
                        180000L,
                    ),
                    0L,
                ),
                180000L,
            )

        val action = intent.action!!

        when (action) {
            "com.google.android.ambientindication.action.AMBIENT_INDICATION_HIDE" -> {
                alarmManager.cancel(mHideIndicationListener)
                ambientIndicationInteractor.hideAmbientMusic()
                Log.i(TAG, "Hiding ambient indication.")
            }
            "com.google.android.ambientindication.action.AMBIENT_INDICATION_SHOW" -> {
                val skipUnlock =
                    intent.getBooleanExtra(
                        "com.google.android.ambientindication.extra.SKIP_UNLOCK",
                        false,
                    )
                val iconOverride =
                    intent.getIntExtra(
                        "com.google.android.ambientindication.extra.ICON_OVERRIDE",
                        0,
                    )
                val iconDescription =
                    intent.getStringExtra(
                        "com.google.android.ambientindication.extra.ICON_DESCRIPTION"
                    )
                val useExtendedInteraction =
                    intent.getBooleanExtra(
                        "com.google.android.ambientindication.extra.USE_EXTENDED_INTERACTION",
                        false,
                    )
                Log.i(TAG, "Using extended interaction: $useExtendedInteraction")

                if (useExtendedInteraction) {
                    ambientIndicationInteractor.setAmbientMusic(
                        text,
                        openIntent,
                        favoritingIntent,
                        iconOverride,
                        skipUnlock,
                        iconDescription,
                        ExtendedIndication(
                            songTitle,
                            artistName,
                            intent.getParcelableExtra<PendingIntent>(
                                "com.google.android.ambientindication.extra.EXPAND_INTENT"
                            ),
                            intent.getBooleanExtra(
                                "com.google.android.ambientindication.extra.IS_RECOGNITION_RESULT",
                                false,
                            ),
                            intent.getBooleanExtra(
                                "com.google.android.ambientindication.extra.IS_SONG_SEARCHING",
                                false,
                            ),
                            null,
                        ),
                    )
                } else {
                    ambientIndicationInteractor.setAmbientMusic(
                        text,
                        openIntent,
                        favoritingIntent,
                        iconOverride,
                        skipUnlock,
                        iconDescription,
                        null,
                    )
                }
                alarmManager.setExact(
                    AlarmManager.ELAPSED_REALTIME_WAKEUP,
                    SystemClock.elapsedRealtime() + ttlMillis,
                    "AmbientIndication",
                    mHideIndicationListener,
                    null,
                )
                Log.i(TAG, "Showing ambient indication.")
            }
            "com.google.android.ambientindication.action.UPDATE_QUICK_AFFORDANCE_STATE" -> {
                val isEnabled =
                    intent.getBooleanExtra(
                        "com.google.android.ambientindication.extra.IS_ENABLED",
                        false,
                    )
                val isActive =
                    intent.getBooleanExtra(
                        "com.google.android.ambientindication.extra.IS_ACTIVE",
                        false,
                    )
                Log.d(
                    TAG,
                    "Received ACTION_UPDATE_QUICK_AFFORDANCE_STATE: isEnabled=$isEnabled, isActive=$isActive",
                )
                val status = AmbientIndicationMusicStatus(isEnabled, isActive)
                ambientIndicationInteractor.ambientIndicationRepository.ambientMusicStatus.value =
                    status
            }
            "com.google.android.ambientindication.action.AMBIENT_INDICATION_EXPAND" -> {
                val albumArtUriString =
                    intent.getStringExtra(
                        "com.google.android.ambientindication.extra.ALBUM_ART_URI"
                    )
                val albumArtUri =
                    if (TextUtils.isEmpty(albumArtUriString)) null else Uri.parse(albumArtUriString)
                val isFavorite =
                    intent.getBooleanExtra(
                        "com.google.android.ambientindication.extra.IS_FAVORITE",
                        false,
                    )
                val dmpIntent =
                    intent.getParcelableExtra<PendingIntent>(
                        "com.google.android.ambientindication.extra.DMP_INTENT"
                    )
                val dmpPackageName =
                    intent.getStringExtra(
                        "com.google.android.ambientindication.extra.DMP_PACKAGE_NAME"
                    )

                val expandedIndicationData =
                    ExpandedIndicationData(
                        dmpIntent = dmpIntent,
                        dmpPackageName = dmpPackageName,
                        albumArtUri = albumArtUri,
                        isFavorite = isFavorite,
                    )

                ambientIndicationInteractor.setAmbientMusic(
                    text,
                    openIntent,
                    favoritingIntent,
                    0,
                    false,
                    "",
                    ExtendedIndication(
                        songTitle,
                        artistName,
                        null,
                        true,
                        false,
                        expandedIndicationData,
                    ),
                )
                alarmManager.setExact(
                    AlarmManager.ELAPSED_REALTIME_WAKEUP,
                    SystemClock.elapsedRealtime() + ttlMillis,
                    "AmbientIndication",
                    mHideIndicationListener,
                    null,
                )
                Log.i(TAG, "Showing expanded ambient indication.")
            }
        }
    }

    fun onUserSwitched() {
        ambientIndicationInteractor.hideAmbientMusic()
    }
}
